File: DsaWithCpp/Tree/BinarySearchTree/IsBST.cpp
#include <iostream>
#include <climits>
#include "Node.cpp"
using namespace std;

bool validBST(Node *root, long min, long max)
{
    if (root == NULL)
        return true;
    if (root->data <= min || root->data >= max)
        return false;
    return (validBST(root->left, min, root->data) &&
            validBST(root->right, root->data, max));
}

bool isBST(Node *root)
{
    return validBST(root, LONG_MIN, LONG_MAX);
}

File: dsa_with_python/tree/binary_search_tree/is_bst.py
import sys
from node import Node


def is_bst(root: Node) -> bool:
    def __is_bst(root: Node, min, max):
        if not root:
            return True
        if root.data <= min or root.data >= max:
            return False
        return __is_bst(root.left, min, root.data) and __is_bst(
            root.right, root.data, max
        )

    return __is_bst(root, -sys.maxsize, sys.maxsize)
